xclib/evaluation/xc_metrics.py

```diff
@@ -616,3 +615,134 @@ def __init__(self, true_labels, inv_propensity_scores=None,
+    """
+    Compute precision@k for values 1...k, faster than using `precision`
+    Arguments:
+    ----------
+    X_indices: np.ndarray
+        2D numpy array with indices sorted in descending order according to score for each row
+    true_labels: csr_matrix
+        ground truth in sparse format
+    k: int
+        compute psprecision for ints in [1, k]
+    Returns:
+    -------
+    np.ndarray: precison values
+    """
+    return _fast_precision_with_indices(X_indices, true_labels.indices.astype(np.int64), true_labels.indptr, k)
+
+@nb.njit(parallel=True)
+def _fast_psprecision_with_indices(predicted, true_indices, true_indptr, k, inv_propensities):
+    """predicted: predicted indices sorted by distance
+       assume predicted indices are unique in each row 
+    """
+    m = predicted.shape[0]
+    cum_intersections = np.zeros((m, k), dtype=np.float64)
+    cum_total = np.zeros((m, k), dtype=np.float64)
+    for i in nb.prange(predicted.shape[0]):
+        intersection = in1d(predicted[i][:k], np.unique(true_indices[true_indptr[i]: true_indptr[i + 1]]))
+        total = inv_propensities[predicted[i][:k]]
+        intersection = np.multiply(total, intersection.astype(np.float64))
+        
+        cum_intersections[i] = np.cumsum(intersection)
+        cum_total[i] = np.cumsum(total)
+
+    precision = mean_rows(cum_intersections / cum_total)
+    
+    return precision
+
+def fast_psprecision_with_indices(X_indices, true_labels, inv_propensities, k):
+    """
+    Compute propensity weighted precision@k for values 1...k, faster than using `psprecision`
+    Arguments:
+    ----------
+    X_indices: np.ndarray
+        2D numpy array with indices sorted in descending order according to score for each row
+    true_labels: csr_matrix
+        ground truth in sparse format
+    k: int
+        compute psprecision for ints in [1, k]
+    Returns:
+    -------
+    np.ndarray: psprecison values
+    """
+    return _fast_psprecision_with_indices(X_indices, true_labels.indices.astype(np.int64), true_labels.indptr, k, inv_propensities)
+
+@nb.njit(parallel=True)
+def _fast_recall_at_k(true_labels_indices, true_labels_indptr, pred_labels_data, pred_labels_indices, pred_labels_indptr, top_k):
+    fracs = -1 * np.ones((len(true_labels_indptr) - 1, ), dtype=np.float32)
+    for i in nb.prange(len(true_labels_indptr) - 1):
+        _true_labels = true_labels_indices[true_labels_indptr[i] : true_labels_indptr[i + 1]]
+        _data = pred_labels_data[pred_labels_indptr[i] : pred_labels_indptr[i + 1]]
+        _indices = pred_labels_indices[pred_labels_indptr[i] : pred_labels_indptr[i + 1]]
+        top_inds = np.argsort(_data)[::-1][:top_k]
+        _pred_labels = _indices[top_inds]
+        if(len(_true_labels) > 0):
+            fracs[i] = len(set(_pred_labels).intersection(set(_true_labels))) / len(_true_labels)
+    return np.mean(fracs[fracs != -1])
+
+def fast_recall_k(X, true_labels, k):
```

Comment:
@DeepakSaini119 would it be possible to instead have a version of recall that computes 1..k, I think it would be pretty similar to how you have implemented precision. Given we a set of m documents `mxk` array of `cum_intersections` we would just need to divide this by a `mx1` array which has the number of ground truth for each elt. Am I missing something?

Reply:
I think adding such a thing would be wasteful. Recall is generally required for large values for k. For instance, if R@100 is required, R@99 would be useless in most cases. But for uniformity, we can have a 1...k version of recall as well